File: typescript/tests/test_signer.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  CallbackSigner,
  CliSigner,
  RawKeySigner,
  createSigner,
} from "../src/signer.js";

describe("CallbackSigner", () => {
  it("delegates to callback", () => {
    const sig = new Uint8Array(65).fill(1);
    const signer = new CallbackSigner(() => sig);
    const result = signer.sign(new Uint8Array(32));
    assert.deepEqual(result, sig);
  });
});

describe("createSigner", () => {
  it("creates CliSigner for cli mode", () => {
    const signer = createSigner("cli");
    assert.ok(signer instanceof CliSigner);
  });

  it("throws for raw mode without key", () => {
    const originalKey = process.env.PAYSKILL_KEY;
    delete process.env.PAYSKILL_KEY;
    assert.throws(() => createSigner("raw"), /No key/);
    if (originalKey) process.env.PAYSKILL_KEY = originalKey;
  });

  it("creates RawKeySigner with key", () => {
    const signer = createSigner("raw", { key: "0x" + "ab".repeat(32) });
    assert.ok(signer instanceof RawKeySigner);
  });

  it("creates CallbackSigner for custom mode", () => {
    const signer = createSigner("custom", {
      callback: () => new Uint8Array(65),
    });
    assert.ok(signer instanceof CallbackSigner);
  });

  it("throws for custom mode without callback", () => {
    assert.throws(() => createSigner("custom"), /callback/);
  });
});

File: python/tests/test_signer.py
"""Tests for signer module."""

import pytest

from payskill.signer import (
    CallbackSigner,
    CliSigner,
    RawKeySigner,
    create_signer,
)


class TestCallbackSigner:
    def test_delegates_to_callback(self) -> None:
        sig = b"\x01" * 65
        signer = CallbackSigner(callback=lambda h: sig)
        assert signer.sign(b"\x00" * 32) == sig


class TestCreateSigner:
    def test_cli_mode(self) -> None:
        signer = create_signer("cli")
        assert isinstance(signer, CliSigner)

    def test_raw_mode_no_key(self) -> None:
        with pytest.raises(ValueError, match="No key"):
            create_signer("raw")

    def test_raw_mode_with_key(self) -> None:
        signer = create_signer("raw", key="0x" + "ab" * 32)
        assert isinstance(signer, RawKeySigner)

    def test_custom_mode(self) -> None:
        signer = create_signer("custom", callback=lambda h: b"\x00" * 65)
        assert isinstance(signer, CallbackSigner)

    def test_custom_mode_no_callback(self) -> None:
        with pytest.raises(ValueError, match="callback"):
            create_signer("custom")

    def test_unknown_mode(self) -> None:
        with pytest.raises(ValueError, match="Unknown"):
            create_signer("unknown")
